librz/diff/diff.c

```diff
@@ -0,0 +1,827 @@
+		match = find_longest_match(diff, block);
+		if (!match) {
+			continue;
+		}
+
+		if (match->size > 0) {
+			if (!rz_list_append(matches, match)) {
+				RZ_LOG_ERROR("rz_diff_matches_new: cannot append match into matches\n");
+				free(match);
+				goto rz_diff_matches_new_fail;
+			}
+			if (block->a_low < match->a && block->b_low < match->b) {
+				if (!stack_append_block(stack, block->a_low, match->a, block->b_low, match->b)) {
+					RZ_LOG_ERROR("rz_diff_matches_new: cannot append low block into stack\n");
+					goto rz_diff_matches_new_fail;
+				}
+			}
+			if (match->a + match->size < block->a_hi && match->b + match->size < block->b_hi) {
+				if (!stack_append_block(stack, match->a + match->size, block->a_hi, match->b + match->size, block->b_hi)) {
+					RZ_LOG_ERROR("rz_diff_matches_new: cannot append high block into stack\n");
+					goto rz_diff_matches_new_fail;
+				}
+			}
+		} else {
+			free(match);
+		}
+		free(block);
+	}
+	rz_list_sort(matches, (RzListComparator)cmp_matches);
+
+	adj_a = 0;
+	adj_b = 0;
+	adj_size = 0;
+	rz_list_foreach (matches, it, match) {
+		if ((adj_a + adj_size) == match->a && (adj_b + adj_size) == match->b) {
+			adj_size += match->size;
+		} else {
+			RzDiffMatch *m = adj_size ? match_new(adj_a, adj_b, adj_size) : NULL;
+			if (adj_size && (!m || !rz_list_append(non_adjacent, m))) {
+				RZ_LOG_ERROR("rz_diff_matches_new: cannot append match into non_adjacent\n");
+				free(m);
+				goto rz_diff_matches_new_fail;
+			}
+			adj_a = match->a;
+			adj_b = match->b;
+			adj_size = match->size;
+		}
+	}
+	match = adj_size ? match_new(adj_a, adj_b, adj_size) : NULL;
+	if (adj_size && (!match || !rz_list_append(non_adjacent, match))) {
+		RZ_LOG_ERROR("rz_diff_matches_new: cannot append match into non_adjacent\n");
+		free(match);
+		goto rz_diff_matches_new_fail;
+	}
+
+	match = match_new(diff->a_size, diff->b_size, 0);
+	if (!match || !rz_list_append(non_adjacent, match)) {
+		RZ_LOG_ERROR("rz_diff_matches_new: cannot append match into non_adjacent\n");
+		free(match);
+		goto rz_diff_matches_new_fail;
+	}
+
+	rz_list_free(matches);
+	rz_list_free(stack);
+	return non_adjacent;
+
+rz_diff_matches_new_fail:
+	rz_list_free(non_adjacent);
+	rz_list_free(matches);
+	rz_list_free(stack);
+	return NULL;
+}
+
+static RzDiffOp *opcode_new(RzDiffOpType type, st32 a_beg, st32 a_end, st32 b_beg, st32 b_end) {
+	RzDiffOp *op = RZ_NEW0(RzDiffOp);
+	if (!op) {
+		return NULL;
+	}
+	op->type = type;
+	op->a_beg = a_beg;
+	op->a_end = a_end;
+	op->b_beg = b_beg;
+	op->b_end = b_end;
+	return op;
+}
+
+static void opcode_set(RzDiffOp *op, RzDiffOpType type, st32 a_beg, st32 a_end, st32 b_beg, st32 b_end) {
+	op->type = type;
+	op->a_beg = a_beg;
+	op->a_end = a_end;
+	op->b_beg = b_beg;
+	op->b_end = b_end;
+}
+
+/**
+ * \brief Generates a list of steps needed to go from A to B
+ *
+ * Generates a list of opcodes that are needed to convert A to B.
+ * */
+RZ_API RzList *rz_diff_opcodes_new(RzDiff *diff) {
```

Comment:
Same here.

Reply:
Also, since you marked the type of the element in the list in some function below, please do it in the rest of the functions too.